chainer/functions/normalization/batch_normalization.py

```diff
@@ -55,15 +56,58 @@ def forward(self, inputs):
             self.running_var = xp.zeros_like(gamma)
         self.mode = _BNMode(x, gamma)
 
+        # TODO(niboshi): Refactor calculation of expander and axis into a
+        # function and call it just before they are used.
+
         # expander inserts singleton dimensions to gamma and beta so that they
         # can be broadcasted with x.
         head_ndim = gamma.ndim + 1
         expander = (None, Ellipsis) + (None,) * (x.ndim - head_ndim)
         self.expander = expander
         self.axis = (0,) + tuple(range(head_ndim, x.ndim))
         self.use_cudnn = self.mode.can_use_cudnn(xp)
+        self.use_ideep = self.mode.can_use_ideep()
+
+        if self.use_ideep:
+            # TODO(niboshi): Refactor iDeep part into a separate method
+            expand_dim = False
+            if x.ndim == 2:
+                expand_dim = True
+                x = x[:, :, None, None]
+
+            gamma = gamma[expander]
+            beta = beta[expander]
+            W = numpy.concatenate((gamma, beta), axis=0).reshape((2, -1))
+
+            y, self.mean, self.var, self.inv_std = (
+                intel64.ideep.batchNormalization.Forward(
+                    intel64.ideep.array(x),
+                    intel64.ideep.array(W),
+                    None,
+                    None,
+                    self.eps
+                ))
 
-        if self.use_cudnn:
+            m = x.size // gamma.size
+            adjust = m / max(m - 1., 1.)
+            if (isinstance(self.running_mean, intel64.ideep.mdarray)
```

Comment:
Can we separate this if branch into one for `running_mean` and the other for `running_var`?

Reply:
Fixed.